src/ensembl/src/shared/components/communication-framework/contact-us/contact-us-form/hooks/useSavedForm.ts

```diff
@@ -0,0 +1,94 @@
+/**
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useEffect, useRef, useContext } from 'react';
+
+import IndexedDBContext from 'src/shared/contexts/IndexedDBContext';
+
+type FormFieldType = string | File | File[] | null;
+
+type Form = Record<string, FormFieldType>;
+
+type Params<S> = {
+  formName: string;
+  currentState: S;
+  updateState: (state: S) => void;
+};
+
+type UseSavedForm = <State extends Form>(
+  params: Params<State>
+) => {
+  clearSavedForm: () => void;
+};
+
+const STORE_NAME = 'contact-forms'; // storage used to save data of different contact forms
+
+const useSavedForm: UseSavedForm = (params) => {
+  const { formName, currentState, updateState } = params;
+  const indexedDB = useContext(IndexedDBContext);
+
+  const stateRef = useRef<typeof currentState>(currentState);
+  stateRef.current = currentState; // <-- to be able to save the latest state despite the closures
+
+  useEffect(() => {
+    // recover form state from the storage
+    indexedDB.get(STORE_NAME, formName).then((savedState) => {
+      if (savedState) {
+        updateState(savedState);
+      }
+    });
+  }, []);
+
+  useEffect(() => {
+    // save the form before user refreshes or closes the tab
+    window.addEventListener('beforeunload', saveFormState);
+
+    return () => {
+      window.removeEventListener('beforeunload', saveFormState);
+      saveFormState(); // <-- also save the form on every unmount
+    };
+  }, []);
+
+  const isEmptyForm = (state: typeof currentState) => {
+    let isEmpty = true;
+    for (const value of Object.values(state)) {
+      if (Array.isArray(value) && value.length) {
+        isEmpty = false;
+      } else if (value !== null && value !== '') {
```

Comment:
Can't we get rid of this second condition by changing the first condition to:

`Array.isArray(value) && value?.length`

Reply:
We are allowing the value to be:
- null
- string
- array of files

The first condition checks that if the value is an array, it is not empty. The second condition checks that if the value is a string, it is not empty. I could have made the second condition more explicit by writing it `typeof value === 'string' && value.length`. `Array.isArray(value) && value?.length` will return [false](url) for a string.

Reply:
It could just be:
```javascript
const isEmptyForm = (state: typeof currentState) => {
  return Object.values(state).some(value => value?.length);
}

```

Reply:
I forgot to mention one other possible case: a field value can, potentially, also be of type `File`. Not in the current form — its `files` filed will always be a File array; but who knows what'll come in the future. And the `File` type does not have a `length` property.